src/base/misc/fatfs.c

```diff
@@ -399,12 +425,30 @@ static void set_geometry(fatfs_t *f, unsigned char *b)
   bpb->sectors_per_track = f->secs_track;
   bpb->num_heads = f->heads;
 
-  if (bpb->num_sectors_small != 0) { // Could be any version BPB, assume v3.0
-    // Not compatible with 3.2, or 3.31 & later, but replicates earlier code
-    bpb->v300.hidden_sectors = f->hidden_secs;
-  } else {                          // Must be FAT16B so can assume v3.31+ BPB
-    bpb->v331.hidden_sectors = f->hidden_secs;;
-    bpb->v331.num_sectors_large = f->total_secs;
+  if (BPB_IS_V400(bpb)) {
+    bpb->v400.hidden_sectors = f->hidden_secs;;
+    bpb->v400.num_sectors_large = (f->total_secs < 65536L) ? 0 : f->total_secs;
+    bpb->v400.drive_number = f->drive_num;
+
+  } else if (BPB_IS_V340(bpb)) {
+    bpb->v340.hidden_sectors = f->hidden_secs;;
+    bpb->v340.num_sectors_large = (f->total_secs < 65536L) ? 0 : f->total_secs;
+    bpb->v340.drive_number = f->drive_num;
+
+  } else if (f->sys_type == OLDDRD_D) { // DR-DOS 3.40 / 3.41 has v3.20 BPB
+    bpb->v320.hidden_sectors = f->hidden_secs;
+    bpb->v320.num_sectors_large = (f->total_secs < 65536L) ? 0 : f->total_secs;
+    // drive_number not passed in boot block as string data in 0x1fd position
```

Comment:
My feeling is that it's not passed, it's discovered in the boot code somehow. After all it's only the drive number that was correct at the time the file system was made. It would seem disassembling the real boot block, or perhaps doing format /s on D: under DR-DOS 3.41 and comparing is the only way to find out.